lightning/src/util/persist.rs

```diff
@@ -194,3 +199,800 @@ where
+                // The first is an opaque identifier for this call (used for calling back write completion). The second
+                // is the channel update sequence number.
+                let ns = match MonitorName::try_from(funding_txo) {
+                    Ok(monitor_name) => self.monitor_update_namespace(&monitor_name),
+                    Err(_) => return chain::ChannelMonitorUpdateStatus::PermanentFailure,
+                };
+                let key = match UpdateName::try_from(update.update_id) {
+                    Ok(update_name) => update_name.0,
+                    Err(_) => return chain::ChannelMonitorUpdateStatus::PermanentFailure,
+                };
+                match self.kv.write(&ns, &key, &update.encode()) {
+                    Ok(()) => chain::ChannelMonitorUpdateStatus::Completed,
+                    Err(_) => chain::ChannelMonitorUpdateStatus::PermanentFailure,
+                }
+            }
+            // A new block. Now we need to persist the entire new monitor and discard the old
+            // updates.
+            None => self.persist_new_channel(funding_txo, monitor, update_id),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::chain::chainmonitor::Persist;
+    use crate::chain::ChannelMonitorUpdateStatus;
+    use crate::events::{ClosureReason, MessageSendEventsProvider};
+    use crate::ln::functional_test_utils::*;
+    use crate::util::logger::{Level, Record};
+    use crate::util::test_utils;
+    use crate::{check_added_monitors, check_closed_broadcast};
+    //use rand::{distributions::Alphanumeric, thread_rng, Rng};
+    use std::{collections::HashMap, io::Cursor, sync::Mutex};
+
+    // =================================
+    // TEST UTILITIES
+    // =================================
+
+    pub struct StdoutLogger {
+        level: Level,
+    }
+
+    impl StdoutLogger {
+        pub fn new(log_level: Level) -> Self {
+            Self { level: log_level }
+        }
+    }
+
+    impl Logger for StdoutLogger {
+        fn log(&self, record: &Record) {
+            if record.level >= self.level {
+                let raw_log = record.args.to_string();
+                let log = format!(
+                    "{:?} {:<5} [{}:{}] {}\n",
+                    // Note that a "real" lightning node almost certainly does *not* want subsecond
+                    // precision for message-receipt information as it makes log entries a target for
+                    // deanonymization attacks. For testing, however, its quite useful.
+                    std::time::SystemTime::now(),
+                    record.level.to_string(),
+                    record.module_path,
+                    record.line,
+                    raw_log
+                );
+                print!("{}", log);
+            }
+        }
+    }
+
+    struct HashMapKVStore {
```

Comment:
I added the read-only mode, simple enough.

But the "with real monitors" test fails now. It's interesting, because the panic is in the `check_persisted_data!` macro from `test_utils`.

```
TRACE node 0 [lightning::ln::channelmanager : lightning/src/ln/channelmanager.rs, 3108] Generating channel update for channel 74c52ab4f11296d62b66a6dba9513b04a3e7fb5a09a30cee22fce7294ab55b7e
DEBUG node 0 [lightning::ln::channelmanager : lightning/src/ln/channelmanager.rs, 6115] Received channel_update for channel 74c52ab4f11296d62b66a6dba9513b04a3e7fb5a09a30cee22fce7294ab55b7e.
DEBUG node 1 [lightning::ln::channelmanager : lightning/src/ln/channelmanager.rs, 6115] Received channel_update for channel 74c52ab4f11296d62b66a6dba9513b04a3e7fb5a09a30cee22fce7294ab55b7e.
thread 'util::persist::tests::test_persister_with_real_monitors' panicked at 'called `Result::unwrap()` on an `Err` value: Custom { kind: InvalidData, error: "Nonsense bytes didn't map to the type they were interpreted as, context: 7e5bb54a29e7fc22ee0ca3095afbe7a3043b51a9dba6662bd69612f1b42ac574_0'" }', lightning/src/util/persist.rs:800:9
```

Any thoughts on what might be going on here? I'm wondering if it has something to do with the `Reader` type I was using vs. the one you're using in `TestStore`?

Reply:
I added the read-only mode, simple enough.

Disregard earlier comment -- the macro the test is breaking on is defined in the test, I thought it was imported from elsewhere.

Reply:
@tnull, would appreciate your 👀 on this branch at this point. I am a little baffled as to why the `test_persister_with_real_monitors` test would fail with the `TestStore`. 

The failure seems to be in the `check_persisted_data` macro (which I copy-pasted from another test), where it `unwrap`s the `read_channelmonitors` result. Which is pretty interesting because that, in turn, is reading through the `ReadableArgs` trait and not anything particularly novel, AFAICT. 

```
thread 'util::persist::tests::test_persister_with_real_monitors' panicked at lightning/src/util/persist.rs:802:9:
called `Result::unwrap()` on an `Err` value: Custom { kind: InvalidData, error: "MonitorDecodeFailed, Nonsense bytes didn't map to the type they were interpreted as, context: 7e5bb54a29e7fc22ee0ca3095afbe7a3043b51a9dba6662bd69612f1b42ac574_0'" }
```

Reply:
Rebased, and the test is back to working.